docs/developers/developer-reference/smart-contract-build-reference.md

```diff
@@ -0,0 +1,324 @@
+
+If we are not careful, there is a risk of adding unwanted endpoints to the contract. A classic example is when we have a crate with multiple modules, of which only one is imported into the smart contract. In some older versions you might have gotten unwanted endpoints from the other modules of that crate. In order to avoid this, we are using the ABI to generate a curated list of endpoints in each `wasm` crate. This way, our contracts always have the exact same endpoints as the ones specified in the ABIs.
+
+This requires code generation. The `meta` crate will handle this code generation too. An example of such generated code lies below:
+
+```rust
+// Code generated by the elrond-wasm multi-contract system. DO NOT EDIT.
+
+////////////////////////////////////////////////////
+////////////////// AUTO-GENERATED //////////////////
+////////////////////////////////////////////////////
+
+// Number of endpoints: 2
+
+#![no_std]
+
+elrond_wasm_node::wasm_endpoints! {
+    adder
+    (
+        getSum
+        add
+    )
+}
+
+elrond_wasm_node::wasm_empty_callback! {}
+```
+
+The `elrond_wasm_node` macros help keep even this generated code to a minimum.
+
+For multi-contract builds, one `wasm` crate needs to be generated for each of the output contracts:
+- The main wasm crate must always reside in the `wasm` folder. The source file is auto-generated, but the `Cargo.toml` must be provided by the developer.
+- The other wasm contracts (called "secondary") receive a crate folder starting with `wasm-`, e.g. `wasm-multisig-view`. These crates are fully generated based on data from `multicontract.toml`. The respective `Cargo.toml` files are based on the the `Cargo.toml` of the main wasm crate. All configs are taken from there, except for the crate name.
+- Warning: Any folders starting with `wasm-` that are unaccounted for will be removed without prompt. This is to keep the folder structure clean in case of renames.
+
+
+### e. Meta crate: the actual WASM build
+
+The previous two steps happen by just calling `cargo run` in the meta crate, but to actually perform a build, one must call `cargo run build`.
+
+With the ABI information and the code generated, the meta crate can now actually build all the WASM contracts, one for each output contract.
+
+The optimizer `wasm-opt` is also run on each of the resulting binaries (unless opted-out in the configs).
+
+The rust compiler places the result in the designated `target` folder, but for convenience, the meta crate moves the executables to the contract's `output` folder and renames them according to the configured names.
+
+You might have performed this step automatically from erdpy, but erdpy actually just calls the meta crate to do this job. This is because at this point only the meta crate has access to the ABIs and can do it that easily.
```

Comment:
```suggestion
You might have performed this step automatically from erdpy, but erdpy actually calls the meta crate to do this job. This is because at this point only the meta crate has access to the ABIs and can do it that easily.
```